### PR TITLE
Feat/user register backend

### PR DESCRIPTION
**Pull request by muratcanakcay**
New user registration implemented.
Project structure improved.

### CODE REVIEW

**Review by wolftender — APPROVED** [review]
looks fine, left a comment but the code is ok

File: backend/appsettings.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
   },
 
   "Db": {
-    "ConnectionString": "Host=localhost:5432;Username=admin;Database=CbtDb"
+    "ConnectionString": "Host=localhost:5432;Username=admin;Password=admin;Database=CbtDb"
```
[the remark this answers is not in the document]
Comment (muratcanakcay) [reply backend/appsettings.json:17]: yes we can do that. we can put a config file push to repo and then gitignore it in local dev. this way repo uses the pushed file and we use our own configs.

File: backend/Services/UserService.cs [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
-    // basic operations
-    Task<User?> GetUserById(int id);
-    Task<User?> GetUserByEmail(string email);
+public class AuthenticationCredentialsException : Exception { }
 
-    // authentication
-    Task<UserAuthenticationResponse> Authenticate(UserAuthenticationRequest model);
+public class RegistrationException : Exception {
+    public new string Message { get; set; } = default!;
+    public RegistrationException() { }
```
Comment (shilangyu) [inline backend/Services/UserService.cs]: This violates nullability

File: backend/Services/UserService.cs [hunk 3]
```diff
@@ -26,17 +27,74 @@ public class UserService : IUserService {
         this.dbContext = dbContext;
     }
 
-    public Task<User?> GetUserById(int id) {
-        return dbContext.Users.SingleOrDefaultAsync(e => e.Id == id);
+    public async Task<User?> GetUserByIdAsync(int userId) {
+        return await dbContext.Users.SingleOrDefaultAsync(e => e.Id == userId);
+    }
```
Comment (shilangyu) [inline backend/Services/UserService.cs:31]: async/await not needed

File: backend/Services/UserService.cs [hunk 4]
```diff
@@ -26,17 +27,74 @@ public class UserService : IUserService {
...
+
+    public async Task<User?> GetUserByEmailAsync(string email) {
+        return await dbContext.Users.SingleOrDefaultAsync(e => e.Email == email);
+    }
```
Comment (shilangyu) [inline backend/Services/UserService.cs:35]: ditto

File: backend/Services/UserService.cs [hunk 5]
```diff
@@ -26,17 +27,74 @@ public class UserService : IUserService {
...
+
+    public async Task<List<User>> GetAllUsersAsync() {
+        return await dbContext.Users.ToListAsync();
+    }
```
Comment (shilangyu) [inline backend/Services/UserService.cs:39]: ditto

**Review by shilangyu — APPROVED** [review]

**Pull request merged**